File: test/checkpoint.test.ts
import { describe, expect, it } from "vitest";
import { messagesAfterCheckpoint } from "../src/checkpoint.js";
import type { BeeperMessage } from "../src/types.js";

const message = (id: string, isSender = false): BeeperMessage => ({ id, chatID: "chat", isSender, type: "TEXT" });

describe("messagesAfterCheckpoint", () => {
  it("bootstraps at the latest incoming message without replaying history", () => {
    expect(messagesAfterCheckpoint([message("3"), message("2", true), message("1")])).toMatchObject({
      messages: [], nextCheckpoint: "3", bootstrapped: true,
    });
  });

  it("returns only new incoming messages, oldest first, and ignores our messages", () => {
    const result = messagesAfterCheckpoint([message("4"), message("3", true), message("2"), message("1")], "1");
    expect(result.messages.map((item) => item.id)).toEqual(["2", "4"]);
    expect(result.nextCheckpoint).toBe("4");
  });

  it("does not replay a page when a stale checkpoint is missing", () => {
    expect(messagesAfterCheckpoint([message("3"), message("2")], "old")).toMatchObject({
      messages: [], nextCheckpoint: "3", checkpointFound: false, bootstrapped: false,
    });
  });
});

File: checkpoint_test.go
package main

import "testing"

func msg(id string, sender bool) BeeperMessage {
	return BeeperMessage{ID: id, ChatID: "chat", IsSender: sender, Type: "TEXT"}
}

func TestCheckpointBootstrapsWithoutHistory(t *testing.T) {
	result := MessagesAfterCheckpoint([]BeeperMessage{msg("3", false), msg("2", true), msg("1", false)}, "")
	if !result.Bootstrapped || result.NextCheckpoint != "3" || len(result.Messages) != 0 {
		t.Fatalf("unexpected result: %#v", result)
	}
}

func TestCheckpointReturnsIncomingOldestFirst(t *testing.T) {
	result := MessagesAfterCheckpoint([]BeeperMessage{msg("4", false), msg("3", true), msg("2", false), msg("1", false)}, "1")
	if len(result.Messages) != 2 || result.Messages[0].ID != "2" || result.Messages[1].ID != "4" || result.NextCheckpoint != "4" {
		t.Fatalf("unexpected result: %#v", result)
	}
}

func TestCheckpointDoesNotReplayWhenMissing(t *testing.T) {
	result := MessagesAfterCheckpoint([]BeeperMessage{msg("3", false), msg("2", false)}, "old")
	if result.CheckpointFound || result.Bootstrapped || len(result.Messages) != 0 || result.NextCheckpoint != "3" {
		t.Fatalf("unexpected result: %#v", result)
	}
}
